ompi/mca/part/persist/part_persist.c

```diff
@@ -39,6 +47,492 @@ ompi_part_persist_t ompi_part_persist = {
+    request->req_ompi.req_status.MPI_ERROR = OMPI_SUCCESS;
+    request->req_part_complete = true;
+    ompi_request_complete(&(request->req_ompi), true );
+}
+
+/**
+ * mca_part_persist_progress is the progress function that will be registered. It handles
+ * both send and recv request testing and completion. It also handles freeing requests,
+ * after MPI_Free is called and the requests have become inactive.
+ */
+static int
+mca_part_persist_progress(void)
+{
+    mca_part_persist_list_t *current;
+    int err;
+    size_t i;
+
+    /* prevent re-entry, */
+    int block_entry = opal_atomic_add_fetch_32(&(ompi_part_persist.block_entry), 1);
+    if(1 < block_entry)
+    {
+        block_entry = opal_atomic_add_fetch_32(&(ompi_part_persist.block_entry), -1);
+        return OMPI_SUCCESS;
+    }
+
+    OPAL_THREAD_LOCK(&ompi_part_persist.lock);
+
+    mca_part_persist_request_t* to_delete = NULL;
+
+    /* Don't do anything till a function in the module is called. */
+    if(-1 == ompi_part_persist.init_world)
+    {
+        OPAL_THREAD_UNLOCK(&ompi_part_persist.lock);
+        block_entry = opal_atomic_add_fetch_32(&(ompi_part_persist.block_entry), -1);
+        return OMPI_SUCCESS;
+    }
+
+    /* Can't do anything if we don't have world */
+    if(0 == ompi_part_persist.init_world) {
+        ompi_part_persist.my_world_rank = ompi_comm_rank(&ompi_mpi_comm_world.comm);
+        err = ompi_comm_idup(&ompi_mpi_comm_world.comm, &ompi_part_persist.part_comm, &ompi_part_persist.part_comm_req);
+        if(err != OMPI_SUCCESS) {
+             exit(-1);
+        }
+        ompi_part_persist.part_comm_ready = 0;
+        err = ompi_comm_idup(&ompi_mpi_comm_world.comm, &ompi_part_persist.part_comm_setup, &ompi_part_persist.part_comm_sreq);
+        if(err != OMPI_SUCCESS) {
+            exit(-1);
+        }
+        ompi_part_persist.part_comm_sready = 0;
+        ompi_part_persist.init_world = 1;
+
+        OPAL_THREAD_UNLOCK(&ompi_part_persist.lock);
+        block_entry = opal_atomic_add_fetch_32(&(ompi_part_persist.block_entry), -1);
+        return OMPI_SUCCESS;
+    }
+
+    /* Check to see if Comms are setup */
+    if(0 == ompi_part_persist.init_comms) {
+        if(0 == ompi_part_persist.part_comm_ready) {
+            ompi_request_test(&ompi_part_persist.part_comm_req, &ompi_part_persist.part_comm_ready, MPI_STATUS_IGNORE);
+        }
+        if(0 == ompi_part_persist.part_comm_sready) {
+            ompi_request_test(&ompi_part_persist.part_comm_sreq, &ompi_part_persist.part_comm_sready, MPI_STATUS_IGNORE);
+        }
+        if(0 != ompi_part_persist.part_comm_ready && 0 != ompi_part_persist.part_comm_sready) {
+            ompi_part_persist.init_comms = 1;
+        }
+        OPAL_THREAD_UNLOCK(&ompi_part_persist.lock);
+        block_entry = opal_atomic_add_fetch_32(&(ompi_part_persist.block_entry), -1);
+        return OMPI_SUCCESS;
+    }
+
+    OPAL_LIST_FOREACH(current, ompi_part_persist.progress_list, mca_part_persist_list_t) {
```

Comment:
Everything before this FOREACH should be called only once per process. And checking for this should definitely not be part of each progress call !!!